tsdb/chunkenc/histogram_meta.go

```diff
@@ -140,6 +175,52 @@ func readZeroThreshold(br *bstreamReader) (float64, error) {
 	}
 }
 
+// isWholeWhenMultiplied checks to see if the number when multiplied by 1000 can
+// be converted into an integer without losing precision.
+func isWholeWhenMultiplied(in float64) bool {
+	i := uint(math.Round(in * 1000))
+	out := float64(i) / 1000
+	return in == out
+}
+
+// putCustomBound writes the custom bound to the bstream. It stores values from 0 to
+// 16.382 (inclusive) that are multiples of 0.001 in an unsigned var int of up to 2 bytes,
+// but needs 1 bit + 8 bytes for other values like negative numbers, numbers greater than
+// 16.382, or numbers that are not a multiple of 0.001, on the assumption that they are
+// less common. In detail:
+//   - Multiply the bound by 1000, without rounding.
+//   - If the multiplied bound is >= 0, <= 16382 and a whole number, store it as an
+//     unsigned var int.
+//   - Otherwise, store 0 as an unsigned var int, followed by the 8 bytes of the original
+//     bound as a float64.
+func putCustomBound(b *bstream, f float64) {
+	tf := f * 1000
+	if tf < 0 || tf > 16382 || !isWholeWhenMultiplied(f) {
+		b.putUvarint(0)
+		b.writeBits(math.Float64bits(f), 64)
+		return
+	}
+	b.putUvarint(uint64(math.Round(tf) + 1))
+}
+
+// readCustomBound reads the custom bound written with putCustomBound.
+func readCustomBound(br *bstreamReader) (float64, error) {
+	b, err := br.readUvarint()
```

Comment:
nit: optimization: technically we could get away with reading a single bit. See comment on line 199

Reply:
noted, let's leave the optimisations for a later PR as suggested